src/vos/vos_pool.c

```diff
@@ -155,31 +155,25 @@ vos_wal_reserve(struct umem_store *store, uint64_t *tx_id)
 	return bio_wal_reserve(store->stor_priv, tx_id);
 }
 
-static void
-vos_wal_on_commit(struct umem_store *store, uint64_t id)
-{
-	struct chkpt_ctx *ctx = store->chkpt_ctx;
-
-	if (ctx == NULL)
-		return; /** Checkpointing not active */
-
-	ctx->cc_commit_id = id;
-
-	if (store->stor_ops->so_wal_id_cmp(store, id, ctx->cc_wait_id) >= 0) {
-		ctx->cc_wake_fn(ctx);
-		store->chkpt_ctx = NULL;
-	}
-}
-
 static inline int
 vos_wal_commit(struct umem_store *store, struct umem_wal_tx *wal_tx, void *data_iod)
 {
 	int rc;
+	struct vos_pool    *pool;
+	struct bio_wal_info wal_info;
 
 	D_ASSERT(store && store->stor_priv != NULL);
 	rc = bio_wal_commit(store->stor_priv, wal_tx, data_iod);
 
-	vos_wal_on_commit(store, wal_tx->utx_id);
+	pool = store->vos_priv;
+
+	if (pool == NULL)
+		return rc;
+
+	bio_wal_query(store->stor_priv, &wal_info);
+
+	pool->vp_update_cb(pool->vp_chkpt_arg, wal_info.wi_ckp_id, wal_info.wi_used_blks,
+			   wal_info.wi_tot_blks);
```

Comment:
In theory, the proper place to wakeup sleeping checkpoint (on WAL space pressure) should be in vos_wal_reserve(). (or in this function but before calling the bio_wal_commit()).

A TX consumes small portion of WAL space then yield on WAL commit, so if there are many concurrent TXs (yield on WAL commit), the WAL used space could have exceeded threshold, but the sleeping checkpoint ULT won't be woken up until the first WAL commit completed (bio_wal_commit() called in this function returns).

It might won't cause problem in real world though, since WAL commit usually should be very fast.

Reply:
There are two kinds of wakeups in current vp_update_cb():
1. Wakeup sleeping checkpoint ULT when WAL free space is running short;
2. Wakeup checkpoint ULT which is waiting for tx committed after the tx committed;

My proposal was to do the 1st wakeup in vos_wal_reserve() and do the 2nd wakeup in vos_wal_commit(), if we don't do the 2nd wakeup on tx commit, checkpoint could wait for commit forever.